pkg/genericapiserver/openapi/openapi.go

```diff
@@ -16,213 +16,138 @@ limitations under the License.
+	if _, ok := o.swagger.Definitions[name]; ok {
+		return nil
 	}
-	for _, swaggerProp := range swaggerModel.Properties.List {
-		if _, ok := ret.Properties[swaggerProp.Name]; ok {
-			return ret, fmt.Errorf("Duplicate property in swagger 1.2 spec: %v", swaggerProp.Name)
-		}
-		ret.Properties[swaggerProp.Name], err = buildProperty(swaggerProp)
-		if err != nil {
-			return ret, err
+	if item, ok := (*o.openAPIDefinitions)[name]; ok {
+		o.swagger.Definitions[name] = item.Schema
+		for _, v := range item.Dependencies {
+			if err := o.buildDefinitionRecursively(v); err != nil {
+				return err
+			}
 		}
+	} else {
+		return fmt.Errorf("Cannot find model definition for %v.", name)
 	}
-	return ret, nil
+	return nil
 }
 
-// buildProperty converts a swagger 1.2 property to an open API property.
-func buildProperty(swaggerProperty swagger.NamedModelProperty) (openAPIProperty spec.Schema, err error) {
-	if swaggerProperty.Property.Ref != nil {
-		return spec.Schema{
-			SchemaProps: spec.SchemaProps{
-				Ref: spec.MustCreateRef("#/definitions/" + *swaggerProperty.Property.Ref),
-			},
-		}, nil
-	}
-	openAPIProperty = spec.Schema{
-		SchemaProps: spec.SchemaProps{
-			Description: swaggerProperty.Property.Description,
-			Default:     getDefaultValue(swaggerProperty.Property.DefaultValue),
-			Enum:        make([]interface{}, len(swaggerProperty.Property.Enum)),
-		},
-	}
-	for i, e := range swaggerProperty.Property.Enum {
-		openAPIProperty.Enum[i] = e
-	}
-	openAPIProperty.Minimum, err = getFloat64OrNil(swaggerProperty.Property.Minimum)
-	if err != nil {
-		return spec.Schema{}, err
-	}
-	openAPIProperty.Maximum, err = getFloat64OrNil(swaggerProperty.Property.Maximum)
-	if err != nil {
-		return spec.Schema{}, err
-	}
-	if swaggerProperty.Property.UniqueItems != nil {
-		openAPIProperty.UniqueItems = *swaggerProperty.Property.UniqueItems
-	}
-
-	if swaggerProperty.Property.Items != nil {
-		if swaggerProperty.Property.Items.Ref != nil {
-			openAPIProperty.Items = &spec.SchemaOrArray{
-				Schema: &spec.Schema{
-					SchemaProps: spec.SchemaProps{
-						Ref: spec.MustCreateRef("#/definitions/" + *swaggerProperty.Property.Items.Ref),
-					},
-				},
-			}
-		} else {
-			openAPIProperty.Items = &spec.SchemaOrArray{
-				Schema: &spec.Schema{},
-			}
-			openAPIProperty.Items.Schema.Type, openAPIProperty.Items.Schema.Format, err =
-				buildType(swaggerProperty.Property.Items.Type, swaggerProperty.Property.Items.Format)
-			if err != nil {
-				return spec.Schema{}, err
-			}
-		}
+// buildDefinitionForType build a definition for a given type and return a referable name to it's definition.
+// This is the main function that keep track of definitions used in this spec and is depend on code generated
+// by k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen.
+func (o *openAPI) buildDefinitionForType(sample interface{}) (string, error) {
+	t := reflect.TypeOf(sample)
+	if t.Kind() == reflect.Ptr {
```

Comment:
If this gets called once per type on startup then it doesn't matter, but if not then in a followup it might be reasonable to replace this with a generated type switch rather than using reflection. It probably won't get called often enough to matter. I just mention this because we've had to do similar things in the api encode/decode paths.